### ISSUE TITLE
Autoquoting

### ISSUE THREAD

**Issue by angerman** [haskell-nix/hnix#282]
When using `hnix` in [Cabal2Nix](https://github.com/angerman/Cabal2Nix), I have this rather annoying construct
```haskell
quoted :: (IsString a, Semigroup a) => a -> a
quoted str = "\"" <> str <> "\""
```
in [Cabal2Nix.Util](https://github.com/angerman/Cabal2Nix/blob/a3930a5d06a55554f45a23ccdcad8568a0ad1535/lib/Cabal2Nix/Util.hs)

I use this to sanitize various names that might not end up being allowed symbols. Mostly on the LHS of binds.

I wonder if we should turn symbols that are invalid (e.g. `foo.` into `"foo."` or `1.2.3.4` into `"1.2.3.4"`, and ensure that the `NExpr` is at least valid by construction?

**Comment by shlevy** [comment]
I'd hope the pretty-printer does escaping already?

**Comment by Ericson2314** [comment]
Yeah, last I checked I think Nix allows anything but absolutely empty string as an attribute name.

**Comment by angerman** [comment]
Not so much :(

```nix
Welcome to Nix version 2.0. Type :? for help.

nix-repl> { a. = 1; }
error: syntax error, unexpected '=', expecting ID or OR_KW or DOLLAR_CURLY or '"', at (string):1:6

nix-repl> { 3a = 1; }
error: syntax error, unexpected INT, at (string):1:3

nix-repl> { 1.2.3.4 = 1; }
error: syntax error, unexpected FLOAT, at (string):1:3

nix-repl> { a.b.c.d = 1; }
{ a = { ... }; }
```

the last one might be one that ends up with unexpected results.

**Comment by shlevy** [comment]
> nix-repl> { "a." = 1; }
> { "a." = 1; }
>

**Comment by Ericson2314** [comment]
@angerman I mean "semantically". The quote is parsed away:

```nix
{ a = 1; } ? "a"
   ==> true
```

**Comment by angerman** [comment]
@Ericson2314 Ahh ok ;-)

**Comment by Profpatsch** [comment]
@angerman I’m not sure if that’s exactly your use-case, but this commit should go into the next release: https://github.com/jwiegley/hnix/commit/8b4c137a3b125f52bb78039a9d201492032b38e8

**Comment by angerman** [comment]
@Profpatsch cool! Does it do LHS of binds as well?
e.g.
```haskell
mkNonRecSet [
   "3a"      $= mkNull
 , "1.2.3.4" $= mkNull
 , "foo."    $= mkNull
]
```

my use case is mostly in translating `cabal` expressions to `nix` expressions. And `parsec` has some really great `test-suite` name: http://hackage.haskell.org/package/parsec-3.1.13.0/parsec.cabal

Where I end up hand-quoting the key (as I didn't want to add `quoted` for each and every key...)
https://github.com/angerman/hackage.nix/blob/master/9066e9f2ac56648e83d68c7b59022363-parsec/3.1.13.0.nix#L35

**Comment by Profpatsch** [comment]
> Does it do LHS of binds as well?

No, but on first glance it looks to me like the left side of `$=` could benefit from that logic as well. @jwiegley what do you say?

**Comment by jwiegley** [comment]
Sounds reasonable to me.

**Comment by angerman** [comment]
@Profpatsch @jwiegley so who's going to make that change? Should I make that or is that going to be grafted onto jwiegley/hnix@8b4c137 by someone?

**Comment by Profpatsch** [comment]
@angerman please, do. I’ll review it. :)
